emsdk_env.sh

```diff
@@ -1,28 +1,64 @@
 # This script is sourced by the user and uses
-# their shell. Try not to use bashisms.
-
+# their shell.
+#
+# This script tries to find its location but
+# this does not work in every shell.
+#
+# It is known to work in bash, zsh and ksh
+#
 # Do not execute this script without sourcing,
 # because it won't have any effect then.
 # That is, always run this script with
 #
-#     . ./emsdk_env.sh
+#     . /path/to/emsdk_env.sh
+#
 # or
-#     source ./emsdk_env.sh
+#
+#     source /path/to/emsdk_env.sh
 #
 # instead of just plainly running with
 #
 #     ./emsdk_env.sh
 #
 # which won't have any effect.
-if [ -z "$BASH_SOURCE" ]; then
-  if [ ! -f "./emsdk.py" ]; then
-    echo "error: You must be in the same directory as emsdk_env.sh when sourcing it (or switch to the bash shell)" 1>&2
+
+CURRENT_SCRIPT=
+DIR="."
+
+if [ -n "$BASH_SOURCE" ]; then
+  CURRENT_SCRIPT="$BASH_SOURCE"
+elif [ -n "$ZSH_VERSION" ]; then
+  CURRENT_SCRIPT="${(%):-%x}"
+elif [ -n "$KSH_VERSION" ]; then
+  CURRENT_SCRIPT=${.sh.file}
+fi
+
+if [ -n "$CURRENT_SCRIPT" ]; then
+  DIR=$(dirname "$CURRENT_SCRIPT")
+  # now work out actual DIR if this is part of a symlink
```

Comment:
Can you explain a little more why this is needed?  i.e. what would i need to do in order to setup a situation that would need this code.   (Edit: I just saw the added tests which might help explain).


I find it quite hard to follow exactly what this is doing but I guess is basically an implementation of `realpath`?

Reply:
yeah, this is an amalgamation of a lot of code I've found and collected to help get the real path to a script. I've needed it for work more than once.

It basically tries extra hard to find the real emsdk directory whether the directory itself is symlinked elsewhere or the emsdk_env.sh script itself is a symlink.

That's why I had to write some tests to see if I could verify all the combinations.

Reply:
Could you add a comment here to that effect.   If its trying to emulate the `realpath` command perhaps say that, and why we can't use use it.

Reply:
sounds good